internals/osutil/mkdirallchown.go

```diff
@@ -31,16 +40,36 @@ var mu sync.Mutex
 
 // MkdirAllChown is like os.MkdirAll but it calls os.Chown on any
 // directories it creates.
-func MkdirAllChown(path string, perm os.FileMode, uid sys.UserID, gid sys.GroupID) error {
+func MkdirAllChown(path string, perm os.FileMode, flags MkdirFlags, uid sys.UserID, gid sys.GroupID) error {
 	mu.Lock()
 	defer mu.Unlock()
 	if uid == NoChown && gid == NoChown {
-		return os.MkdirAll(path, perm)
+		if flags&MkdirChmod != 0 {
+			return mkdirAllChmod(filepath.Clean(path), perm)
+		} else {
+			return os.MkdirAll(path, perm)
+		}
+	}
+	return mkdirAllChown(filepath.Clean(path), perm, flags, uid, gid)
+}
+
+func mkdirAllChmod(path string, perm os.FileMode) error {
+	if _, err := os.Stat(path); err == nil {
+		return nil
+	}
+
+	parent := filepath.Dir(path)
+	if err := mkdirAllChmod(parent, perm); err != nil {
```

Comment:
This doesn't look like the right logic, or at least not the same logic as `mkdirAllChown`, which uses the `s.IsDir()` to see if it's a directory or not.

Reply:
I see, the `IsDir` is necessary because the path might already exists as a file. I missed this in the first draft. Will change this function to the same logic as `mkdirAllChown`.